File: core/inpainter/schema.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from inpainter.errors import CoreError
from inpainter.paths import schema_dir


def load_schema(name: str) -> dict[str, Any]:
    path = schema_dir() / f"{name}.json"
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except FileNotFoundError as exc:
        raise CoreError(f"missing schema: {name}") from exc
    except json.JSONDecodeError as exc:
        raise CoreError(f"unreadable schema: {name}") from exc
    if not isinstance(data, dict):
        raise CoreError(f"schema is not an object: {name}")
    return data


def apply_defaults(name: str, payload: dict[str, Any] | None) -> dict[str, Any]:
    schema = load_schema(name)
    incoming = payload or {}
    result: dict[str, Any] = {}
    properties = schema.get("properties")
    if not isinstance(properties, dict):
        return dict(incoming)
    for key, spec in properties.items():
        if not isinstance(spec, dict):
            continue
        if key in incoming:
            result[key] = incoming[key]
        else:
            result[key] = spec.get("default")
    return result


def schema_path(name: str) -> Path:
    return schema_dir() / f"{name}.json"

File: core/src/schema.ts
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { CoreError } from "./errors.ts";
import { schemaDir } from "./paths.ts";

type JsonSchema = {
  properties?: Record<string, { default?: unknown }>;
};

export function loadSchema(name: string): JsonSchema {
  const path = join(schemaDir(), `${name}.json`);
  try {
    const data = JSON.parse(readFileSync(path, "utf8")) as unknown;
    if (data === null || typeof data !== "object" || Array.isArray(data)) {
      throw new CoreError(`schema is not an object: ${name}`);
    }
    return data as JsonSchema;
  } catch (error) {
    if (error instanceof CoreError) {
      throw error;
    }
    const err = error as NodeJS.ErrnoException;
    if (err.code === "ENOENT") {
      throw new CoreError(`missing schema: ${name}`);
    }
    throw new CoreError(`unreadable schema: ${name}`);
  }
}

export function applyDefaults(
  name: string,
  payload: Record<string, unknown> | null | undefined,
): Record<string, unknown> {
  const schema = loadSchema(name);
  const incoming = payload ?? {};
  const properties = schema.properties;
  if (!properties || typeof properties !== "object") {
    return { ...incoming };
  }
  const result: Record<string, unknown> = {};
  for (const [key, spec] of Object.entries(properties)) {
    if (!spec || typeof spec !== "object") {
      continue;
    }
    result[key] = key in incoming ? incoming[key] : spec.default;
  }
  return result;
}
